### DISCUSSION TITLE
Resolution Convergence in Cylindrical Simulations

### DISCUSSION THREAD

**Discussion by msanch23** [NanoComp/meep#2918] [Q&A]
Hello! 

I'm running into an issue with using axisymmetry to simulate a Bullseye cavity. I've validated the cavity in 3D simulations by converging PML thickness, spatial dimensions, simulation time (I use field decay), and resolution. Below is the resolution convergence. 
![image](https://github.com/user-attachments/assets/0e6d2327-f570-4d87-bc05-52dec5c8ed35)
As one can see the simulation converges nicely with increasing resolution to around 690 nm - this simulation takes a couple of hours using multiple cores. 
So, I wanted to take advantage of the axisymmetry of my cavity:
![image](https://github.com/user-attachments/assets/10277374-34e5-4757-a3ee-1b030f1ba0a1)
All the parameters in the axisymmetric simulation were converging nicely until I got to resolution, where I'm encountering a strange pattern that I can't seem to fix. I want to mention that the PML thickness, r, z, and simulation time were all converged using a resolution of 60 pixel/um, and I found that they change the Q moreso than the resonant frequency.
![image](https://github.com/user-attachments/assets/4a02bb16-9d76-48fb-b3b4-1a6a3f072036)
The source I'm using is shifted to 1.5Δr:
```
wl = 0.690
fcen = 1/wl
df = 0.05*fcen

sources = [mp.Source(src=mp.GaussianSource(fcen,fwidth=df),
                     component=mp.Er,
                     center=mp.Vector3(1.5/resolution,0,0))]
```
and I'm using m=1 along with Harminv to find the cavity's resonance and Q:
```
sim = mp.Simulation(resolution=resolution,
                        cell_size=cell_size,
                        boundary_layers=pml_layers,
                        sources=sources,
                        geometry=geometry,
                        dimensions=mp.CYLINDRICAL,
                        m=1,
                        default_material=mp.air)

h = mp.Harminv(mp.Er, mp.Vector3(1.5/resolution,0,0), fcen, df)

sim.run(mp.after_sources(h),
        until_after_sources=mp.stop_when_fields_decayed(10,
                                                        mp.Er,
                                                        mp.Vector3(1.5/resolution,0,0),
                                                        1e-05))
```
Any ideas what might be the issue? I would appreciate any and all feedback! Thank you.

**Comment by smartalecH** [comment]
Hmm so you have to be careful using `stop_when_fields_decayed` with resonant problems, especially with a tolerance of 10^-5, which is relatively high. 

You can try dropping the tolerance quite a bit. But better yet, just prescribe a _fixed_ time where you get enough periods in (this should make your spatial resolution convergence plot more consistent and prevent the mode hopping). You may need to play with that R PML a bit more too, since that's only a quasi-PML. 

Also, I know you're just trying to resolve the convergence issues for now. But once you're ready to simulate an actual dipole, be sure to check out how to properly to do so in cylindrical coordinates: https://meep.readthedocs.io/en/master/Python_Tutorials/Cylindrical_Coordinates/#nonaxisymmetric-dipole-sources
